File: detox/ios/Detox/Utilities/ViewHierarchyGenerator.swift
//
//  ViewHierarchyGenerator.swift
//  Detox
//
//  Created by Georgy Steshin on 18/08/2024.
//  Copyright © 2024 Wix. All rights reserved.
//

import Foundation
import UIKit
import WebKit

private let GET_HTML_SCRIPT = """
    (function() {
    const blacklistedTags = ['script', 'style', 'head', 'meta'];
    const blackListedTagsSelector = blacklistedTags.join(',');

    // Clone the entire document
    var clonedDoc = document.documentElement.cloneNode(true);

    // Remove all <script> and <style> tags from the cloned document
    var scripts = clonedDoc.querySelectorAll(blackListedTagsSelector);
    scripts.forEach(function(script) {
        script.remove();
    });

    // Create an instance of XMLSerializer
    var serializer = new XMLSerializer();

    // Serialize the cloned DOM to a string
    var serializedHtml = serializer.serializeToString(clonedDoc);

    // Return the serialized HTML as a string
    return serializedHtml;
})();
"""

struct ViewHierarchyGenerator {
    private static let maxDepth = 200

    @MainActor
    static func generateXml(injectingAccessibilityIdentifiers shouldInject: Bool) async -> String {
        let xmlHeader = "<?xml version=\"1.0\" encoding=\"utf-8\"?>"
        let windows = UIWindow.allKeyWindowSceneWindows
            .filter { !isVisualizerWindow($0) }

        var viewHierarchy = ""
        for window in windows {
            viewHierarchy += await generateXmlForViewHierarchy(window, depth: 1, indexPath: [], shouldInjectIdentifiers: shouldInject)
        }

        return """
        \(xmlHeader)
        <ViewHierarchy>\(viewHierarchy)
        </ViewHierarchy>
        """
    }

    private static func isVisualizerWindow(_ window: UIWindow) -> Bool {
        return NSStringFromClass(type(of: window)) == "DTXTouchVisualizerWindow"
    }

    @MainActor
    private static func generateXmlForViewHierarchy(
        _ view: UIView,
        depth: Int,
        indexPath: [Int],
        shouldInjectIdentifiers: Bool
    ) async -> String {
        guard depth <= maxDepth else { return "" }

        let indent = String(repeating: " ", count: depth)
        let elementName = String(describing: type(of: view))
        let attributes = generateAttributes(for: view, indexPath: indexPath, shouldInjectIdentifiers: shouldInjectIdentifiers)

        var xml = "\n\(indent)<\(elementName)\(attributes)"

        if let webView = view as? WKWebView {
            let htmlContent = await getHtmlFromWebView(webView)
            xml += ">\n\(indent)\t<![CDATA[\(htmlContent)]]>"
            xml += "\n\(indent)</\(elementName)>"
        } else if view.subviews.isEmpty {
            xml += " />"
        } else {
            xml += ">"
            for (index, subview) in view.subviews.enumerated() {
                let subviewIndexPath = indexPath + [index]
                xml += await generateXmlForViewHierarchy(
                    subview,
                    depth: depth + 1,
                    indexPath: subviewIndexPath,
                    shouldInjectIdentifiers: shouldInjectIdentifiers
                )
            }
            xml += "\n\(indent)</\(elementName)>"
        }

        return xml
    }

    @MainActor
    private static func getHtmlFromWebView(_ webView: WKWebView) async -> String {
        let handler = WebViewHandler()

        do {
            try await handler.waitForPageLoad(webView)
        } catch {
            print("Error waiting for page to load: \(error)")
            return "<!-- Error loading page -->"
        }

        return await withCheckedContinuation { continuation in
            webView.evaluateJavaScript(GET_HTML_SCRIPT) { result, error in
                if let html = result as? String {
                    continuation.resume(returning: html)
                } else if let error = error {
                    continuation.resume(returning: "<!-- Error loading HTML: \(error) -->")
                } else {
                    continuation.resume(returning: "<!--- HTML is empty -->")
                }
            }
        }
    }

    private static func generateAttributes(
        for view: UIView,
        indexPath: [Int],
        shouldInjectIdentifiers: Bool
    ) -> String {
        var attributes: [String: String] = [
            "class": "\(type(of: view))",
            "width": "\(toClampedInt(view.frame.size.width))",
            "height": "\(toClampedInt(view.frame.size.height))",
            "visibility": view.isHidden ? "invisible" : "visible",
            "alpha": "\(view.alpha)",
            "focused": "\(view.isFocused)",
            "value": view.accessibilityValue ?? "",
            "label": view.accessibilityLabel ?? ""
        ]

        if view.tag != 0 {
            attributes["tag"] = "\(view.tag)"
        }

        if let superview = view.superview {
            let location = view.convert(view.bounds.origin, to: superview)
            attributes["x"] = "\(toClampedInt(location.x))"
            attributes["y"] = "\(toClampedInt(location.y))"
        }

        if shouldInjectIdentifiers {
            let injectedPrefix = "detox_temp_"
            let injectedIdentifier = "\(injectedPrefix)\(indexPath.map { String($0) }.joined(separator: "_"))"

            if let existingTestID = view.accessibilityIdentifier {
                // Override previously injected identifiers
                if existingTestID.hasPrefix(injectedPrefix) {
                    view.accessibilityIdentifier = injectedIdentifier
                }
            } else {
                view.accessibilityIdentifier = injectedIdentifier
            }
        }

        if let testID = view.accessibilityIdentifier {
            attributes["id"] = testID
        }

        if let textView = view as? UITextView {
            attributes["text"] = textView.text ?? ""
        } else if let label = view as? UILabel {
            attributes["text"] = label.text ?? ""
        }

        return attributes
            .filter { !$0.value.isEmpty }
            .map { " \($0.key)=\"\($0.value)\"" }
            .sorted()
            .joined()
    }
}

@MainActor
class WebViewHandler: NSObject, WKNavigationDelegate {
    private var loadCompletion: ((Result<Void, Error>) -> Void)?

    enum WebViewError: Error {
        case timeout
    }

    func waitForPageLoad(_ webView: WKWebView, timeoutAfter seconds: TimeInterval = 10) async throws {
        webView.navigationDelegate = self

        if !webView.isLoading {
            return
        }

        return try await withCheckedThrowingContinuation { continuation in
            let task = Task {
                try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
                loadCompletion?(.failure(WebViewError.timeout))
            }

            self.loadCompletion = { result in
                task.cancel()
                self.loadCompletion = nil
                continuation.resume(with: result)
            }
        }
    }

    nonisolated func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        Task { @MainActor in
            loadCompletion?(.success(()))
        }
    }

    nonisolated func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        Task { @MainActor in
            loadCompletion?(.failure(error))
        }
    }
}

func toClampedInt(_ value: Double) -> Int {
    return Int(min(max(value, Double(Int.min)), Double(Int.max)))
}

File: detox/android/detox/src/full/java/com/wix/detox/espresso/hierarchy/ViewHierarchyGenerator.kt
package com.wix.detox.espresso.hierarchy

import android.util.Xml
import android.view.View
import android.view.ViewGroup
import android.webkit.WebView
import android.widget.TextView
import com.wix.detox.reactnative.ui.getAccessibilityLabel
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import org.xmlpull.v1.XmlSerializer
import java.io.StringWriter
import kotlin.coroutines.resume


private const val GET_HTML_SCRIPT = """
(function() {
    const blacklistedTags = ['script', 'style', 'head', 'meta'];
    const blackListedTagsSelector = blacklistedTags.join(',');

    // Clone the entire document
    var clonedDoc = document.documentElement.cloneNode(true);

    // Remove all <script> and <style> tags from the cloned document
    var scripts = clonedDoc.querySelectorAll(blackListedTagsSelector);
    scripts.forEach(function(script) {
        script.remove();
    });

    // Create an instance of XMLSerializer
    var serializer = new XMLSerializer();

    // Serialize the cloned DOM to a string
    var serializedHtml = serializer.serializeToString(clonedDoc);

    // Return the serialized HTML as a string
    return serializedHtml;
})();
"""

object ViewHierarchyGenerator {
    @JvmStatic
    fun generateXml(shouldInjectTestIds: Boolean): String {
        return runBlocking {
            val rootViews = RootViewsHelper.getRootViews()
            generateXmlFromViews(rootViews, shouldInjectTestIds)
        }
    }

    private suspend fun generateXmlFromViews(rootViews: List<View?>?, shouldInjectTestIds: Boolean): String {
        return StringWriter().use { writer ->
            val serializer = Xml.newSerializer().apply {
                setOutput(writer)
                startDocument(Xml.Encoding.UTF_8.name, true)
                setFeature("http://xmlpull.org/v1/doc/features.html#indent-output", true)
                startTag("", "ViewHierarchy")
            }

            rootViews?.forEach { rootView ->
                rootView?.let {
                    serializeViewHierarchy(it, serializer, shouldInjectTestIds, emptyList())
                }
            }

            serializer.apply {
                endTag("", "ViewHierarchy")
                endDocument()
            }

            writer.toString()
        }
    }

    private suspend fun serializeViewHierarchy(
        view: View,
        serializer: XmlSerializer,
        shouldInjectTestIds: Boolean,
        indexPath: List<Int>
    ) {
        serializer.startTag("", view.javaClass.simpleName)
        serializeViewAttributes(view, serializer, shouldInjectTestIds, indexPath)

        when (view) {
            is WebView -> serializeWebView(view, serializer)
            is ViewGroup -> serializeViewGroupChildren(view, serializer, shouldInjectTestIds, indexPath)
        }

        serializer.endTag("", view.javaClass.simpleName)
    }

    private suspend fun serializeWebView(
        webView: WebView,
        serializer: XmlSerializer,
    ) {
        val html = getWebViewHtml(webView)
        serializer.cdsect(html)
    }

    private suspend fun getWebViewHtml(webView: WebView): String = withContext(Dispatchers.Main) {
        suspendCancellableCoroutine { cancellableContinuation ->
            webView.evaluateJavascript(GET_HTML_SCRIPT) { html ->
                cancellableContinuation.resume(html.unescapeUnicodeString())
            }
        }
    }

    private fun String.unescapeUnicodeString(): String {
        // Replace all Unicode escape sequences (e.g., \u003C -> <)
        return this
            .replace("\\u003C", "<")
            .replace("\\u003E", ">")
            .replace("\\u0022", "\"")
            .replace("\\u0027", "'")
            .replace("\\u0026", "&")
            .replace("\\u003D", "=")
            .replace("\\u002F", "/")
            .replace("\\n", "\n")
    }

    private suspend fun serializeViewGroupChildren(
        view: ViewGroup,
        serializer: XmlSerializer,
        shouldInjectTestIds: Boolean,
        indexPath: List<Int>
    ) {
        for (i in 0 until view.childCount) {
            serializeViewHierarchy(
                view.getChildAt(i),
                serializer,
                shouldInjectTestIds,
                indexPath + i
            )
        }
    }

    private fun serializeViewAttributes(
        view: View,
        serializer: XmlSerializer,
        shouldInjectTestIds: Boolean,
        indexPath: List<Int>
    ) {
        val attributes = mutableMapOf(
            "class" to view.javaClass.name,
            "width" to view.width.toString(),
            "height" to view.height.toString(),
            "visibility" to view.visibilityToString(),
            "alpha" to view.alpha.toString(),
            "focused" to view.isFocused.toString(),
            "value" to (view.contentDescription?.toString() ?: ""),
            "label" to (view.getAccessibilityLabel()?.toString() ?: "")
        )

        val location = IntArray(2).apply { view.getLocationInWindow(this) }
        attributes["x"] = location[0].toString()
        attributes["y"] = location[1].toString()

        if (view is TextView) {
            attributes["text"] = view.text.toString()
        }

        val currentTestId = view.tag?.toString() ?: ""

        val injectedPrefix = "detox_temp_"
        val isTestIdEmpty = currentTestId.isEmpty()
        val isTestIdInjected = currentTestId.startsWith(injectedPrefix)
        val shouldInjectNewTestId = shouldInjectTestIds && (isTestIdEmpty || isTestIdInjected)

        if (shouldInjectNewTestId) {
            val newTestId = "${injectedPrefix}${indexPath.joinToString("_")}"
            view.tag = newTestId
            attributes["id"] = newTestId
        } else {
            attributes["id"] = currentTestId
        }

        attributes
            .filter { (_, value) ->
                !value.isNullOrEmpty()
            }
            .forEach { (key, value) ->
                serializer.attribute("", key, value)
            }
    }

    private fun View.visibilityToString() = when (visibility) {
        View.VISIBLE -> "visible"
        View.INVISIBLE -> "invisible"
        View.GONE -> "gone"
        else -> "unknown"
    }
}
